### PR TITLE
option to delay dataset .bin numpy buffer mmap creation in megatron

### PR DESCRIPTION
**Pull request by anmolgupt**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/data/language_modeling/megatron/indexed_dataset.py [hunk 1]
```diff
@@ -488,25 +488,33 @@ def __getitem__(self, i):
...
 
-    def _do_init(self, path, skip_warmup=True):
+    def _do_init(self, path, skip_warmup=True, delay_data_mmap=False):
         self._path = path
         self._index = self.Index(index_file_path(self._path), skip_warmup)
 
+        if not delay_data_mmap:
+            self._create_data_mmap(skip_warmup)
+        else:
+            logging.info("    skip creating data numpy buffer of mmap...")
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/indexed_dataset.py:515]: remove whitesapce

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py [hunk 2]
```diff
@@ -261,12 +263,12 @@ def build_dataset(index, name):
...
-def get_indexed_dataset_(data_prefix, data_impl, skip_warmup):
+def get_indexed_dataset_(data_prefix, data_impl, skip_warmup, delay_data_mmap):
     """Build indexed dataset."""
     logging.info(' > building dataset index ...')
 
     start_time = time.time()
-    indexed_dataset = make_indexed_dataset(data_prefix, data_impl, skip_warmup)
+    indexed_dataset = make_indexed_dataset(data_prefix, data_impl, skip_warmup, delay_data_mmap=delay_data_mmap)
     logging.info(' > finished creating indexed dataset in {:4f} ' 'seconds'.format(time.time() - start_time))
     logging.info('    number of documents: {}'.format(indexed_dataset.sizes.shape[0]))
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:273]: These log statements only make sense if delay_data_mmap=False?

Comment (anmolgupt) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:273]: The document, shuffle, and sample index files will be created even with `delay_data_mmap=False`. Only the `.bin` files of the dataset will not be accessed when `delay_data_mmap=True`

File: nemo/collections/nlp/data/language_modeling/megatron/indexed_dataset.py (hunk 1 above)
Comment (anmolgupt) [reply nemo/collections/nlp/data/language_modeling/megatron/indexed_dataset.py:515]: Added this whitespace to align with other print statements: https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/data/language_modeling/megatron/indexed_dataset.py#L511-L515 
I can remove it if it's not needed. Please let me know.

Comment (ericharper) [reply nemo/collections/nlp/data/language_modeling/megatron/indexed_dataset.py:515]: Unless there's a reason for it, please remove. Old statements may have just been copied over.

**Review by ericharper — DISMISSED** [review]
LGTM. Thanks!

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**